### PR TITLE
DataSources: Log async datasource plugin import failures

### LINKED ISSUE #126379: DataSources: Close behavioural divergences between async datasource API and legacy DatasourceSrv
## Background

We're migrating from the legacy synchronous `DatasourceSrv` (`public/app/features/plugins/datasource_srv.ts`) to the new async datasource API in `packages/grafana-runtime/src/services/dataSource/` (`getDataSourceInstanceSettings`, `getDataSourceInstanceSettingsList`, `getDataSourceInstance`, `reloadDataSourceInstanceSettings`, `registerRuntimeDataSourceInstance`).

A parity audit + test sweep confirmed the async API reproduces the bulk of legacy behaviour (see the parity test suite added for `getList` ⟷ `getDataSourceInstanceSettingsList` and the expanded `settings.test.ts` / `dataSource.test.ts`). This issue tracks the **genuine behavioural divergences** that remain. Each item is an independent TODO that can be resolved in its own PR.

## TODOs

- [x] **1. Expression refs in `getDataSourceInstance` don't return the preloaded singleton.**
  Legacy `get()` (`datasource_srv.ts`) short-circuits expression references (`__expr__`, `-100`, `Expression`, or a `DataSourceRef` with those types) and returns the preloaded `expressionDatasource` singleton — it never imports a plugin module. The async `getDataSourceInstance` has no short-circuit: it resolves the `__expr__` settings and then calls the importer, which fails because `expressionInstanceSettings.meta.module === ''`. The runtime package cannot import `app/features/expressions` (layering), so the fix is to (a) add an `isExpressionReference` short-circuit in `getDataSourceInstance` that returns a cached instance, and (b) register the Expression singleton into the plugin cache at boot (`public/app/app.ts`) via a dedicated internal helper (NOT `registerRuntimeDataSourceInstance`, which would throw because `__expr__` settings already exist in the cache). The instance must be stored in the runtime cache so it survives `reloadDataSourceInstanceSettings()`. *Currently latent — no production consumer of `getDataSourceInstance` exists yet.* Marked with an `it.todo` in `dataSource.test.ts`. (Fixed by https://github.com/grafana/grafana/pull/126390)

- [x] **2. No `AppEvents.alertError` on plugin-load failure.**
  Legacy `loadDatasource` emits `AppEvents.alertError` via `appEvents` when a plugin fails to import, then rejects with a sanitized message. The async `loadDataSourceInstance` simply rethrows the raw error. Decide whether the async primitive should emit the global alert (it can via `getAppEvents()` from `@grafana/runtime`) or whether the migration layer / React hooks should own user-facing error presentation. (Fixed by https://github.com/grafana/grafana/pull/126404)

- [x] **3. `pluginId` / meta resolution differs (benign — verify before relying on it).**
  Legacy resolves `pluginId` via `getPluginIdFromDatasourceInstanceType(type, name)`, fetches deduplicated meta via `getDatasourcePluginMeta(pluginId)`, and falls back to `instanceSettings.meta` with a warning on miss. The async path passes `settings.meta` straight to the importer. This is benign for normal datasources (`getPluginIdFromDatasourceInstanceType` returns `''`, so legacy already falls back to `instanceSettings.meta`) and for builtins (which carry the correct `meta.id`). Close once confirmed no edge case (e.g. aliased plugin meta) is missed. (Wrote a test to verify that we have the same behavior in both APIs ✅ )

- [x] **4. Patched legacy-plugin `type` value.**
  For plugins that don't extend `DataSourceApi`, legacy patches `anyInstance.type = pluginId` (resolves to `''` for normal datasources); async sets `anyInstance.type = settings.type` (the real type). The async value is arguably more correct. Confirm nothing depends on the legacy `''`/pluginId value, then close. (can't find anything relying on the '' plugin id ✅ )

- [x] **5. Type-only ref with no match: reject vs. fall back.**
  Legacy `get()` rejects with `'no datasource of type'` when a type-only `DataSourceRef` matches nothing. The async `getDataSourceInstance` falls back to the configured default datasource (and `getDataSourceInstanceSettings` resolves to the Grafana DS, because `findByType` → `applyFilters` always appends Grafana). This is intentional and documented at the settings layer, but the legacy `get()` rejection path is a real difference — decide whether any caller relies on the rejection. (Wrote a test to verify that we have the same behavior in both APIs ✅ )

### PR DESCRIPTION
**Pull request by mckn**
**What is this feature?**

When a datasource plugin module fails to import, the async `getDataSourceInstance` now logs the failure via the runtime logger registry (new `grafana/runtime.plugins.datasource` source) and rethrows the raw error, instead of silently rethrowing with no diagnostics.

**Why do we need this feature?**

This closes behavioural divergence #2 between the new async datasource API and the legacy `DatasourceSrv`: legacy `loadDatasource` emitted an `AppEvents.alertError` toast and a sanitized error on plugin-load failure. Rather than replicate that toast (which has no precedent in `grafana-runtime` and risks double-toasts), this follows the convention already established by the new plugin-meta APIs — structured logging in the primitive, raw errors preserved, and user-facing presentation owned by hooks (`useDataSourceInstance` already exposes `error`).

**Who is this feature for?**

Grafana developers/maintainers migrating callers onto the async datasource API.

**Which issue(s) does this PR fix?**:

Fixes #126379

**Special notes for your reviewer:**

This is stacked on #126390 (base branch `mckn/simplify-expression-ds-settings`) and should be re-targeted to `main` once that merges. Frontend-only, no production consumer of `getDataSourceInstance` exists yet so there is no user-visible change today.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: packages/grafana-runtime/src/services/dataSource/dataSource.ts [hunk 1]
```diff
@@ -80,7 +81,16 @@ async function loadDataSourceInstance(cacheUid: string, settings: DataSourceInst
...
   }
 
-  const dsPlugin = await importDataSourcePlugin(settings.meta);
+  let dsPlugin;
+  try {
+    dsPlugin = await importDataSourcePlugin(settings.meta);
+  } catch (error) {
+    logDataSourceInstanceError(`Failed to import datasource plugin ${settings.name} (${settings.uid})`, error, {
+      pluginId: settings.meta.id,
+      uid: settings.uid,
```
Comment (hugohaggmark) [inline packages/grafana-runtime/src/services/dataSource/dataSource.ts:90]: nit: maybe add name as well?
```suggestion
      uid: settings.uid,
      name: settings.name     
```

**Review by hugohaggmark — DISMISSED** [review]

File: packages/grafana-runtime/src/services/dataSource/dataSource.ts (hunk 1 above)
Comment (mckn) [reply packages/grafana-runtime/src/services/dataSource/dataSource.ts:90]: Good idea, added 👍

**Review by hugohaggmark — APPROVED** [review]
